### ISSUE TITLE
Improve topo scalability

### ISSUE THREAD

**Issue by sougou** [vitessio/vitess#3997]
## Problem
As the number of Vitess use cases grow, the existing implementation of topo is facing scalability challenges. We have managed to consolidate and unify the API around all the lockservers. However, not all of them can handle the same type of load. There are currently two problems that are becoming acute:
Consul cannot handle a high QPS of requests: Currently, vtgate polls the lockserver for changes in tablet states. This is achieved by a call to ‘ListDir’ followed by a loop that fetches each tablet record. This, multiplied by the number of vtgates easily overloads a single Consul cluster.
SrvVSchema gets too big: If the number of tablets gets large (>1000), the SrvVSchema file can reach many megabytes, which is something that etcd struggles with.

For problem 1: There is currently a workaround: Issue a ListDir for the tablet ids. But fetch the tablet records only if they are not already present in the existing healthcheck list. The disadvantages of this approach are:

* Tablet type changes will not be captured.
* If a tablet is restarted on a different host, it will not get updated.

This approach is currently implemented under the ‘tablet_refresh_known_tablets’ flag.

For problem 2: We could separate out SrvVSchema into separate files. However, that means we’ll have to watch lots of files, which could be a problem. Also, a separate tree for vschema means that it can become inconsistent with the keyspaces tree.

## Design constraints
1. Restrict to existing set of common operations supported by all lockservers.
2. Lockserver should be used only for initial load, and subsequently only for passive discovery updates. 3. Up-to-date information should be obtained from vttablet health checks. Also, information from vttablet should be treated as authoritative.
4. Do not spam the lockserver.
5. Do not aggregate information from multiple keyspace into a single file.

## Proposal
The following proposed solution attempts to address all of the above problems, while satisfying the specified constraints:

A single journal file is created at the root level of each cell. This file will be human readable (for debuggability). Any event that requires vtgates to update their metadata will cause this file to be updated. The event will be appended at the end of the file.

VTGates will setup a single watch against this file. When they get a notification, they will be able to identify the event that caused the change, and specifically load that information. The entries will contain:

* Event type: add, update or delete.
* Item type: tablet record or vschema.
* Item id: kesypace/shard/tablet type for tablets, and keyspace for vschema
* Timestamp

We’ll setup an expiry time for this journal. Anyone that updates the journal will also expire old items. Journal updates will have to be serialized to ensure no item gets missed. This means that the number of changes per second in the system will eventually become a bottleneck. For now, this is not the case. Compare-and-set or transactional functionality will be used to prevent racing updates from overwriting on each other.

### New tree
With the above scheme, we should be able to move all information related to a keyspace under a single node. In other words, the SrvVSchema will be per-keyspace. When a tablet record or a vschema gets updated, the journal file is also updated. This will cause all vtgates to receive the notification and they will load the specific info that changed.

### Migration
For migration, this essentially boils down to moving the vschema under the keyspace tree, and creating the Journal file.

The move of SrvVSchema can be done through a RebuildVSchemaGraph. Maybe we can look at doing this automatically. For backward compatibility, we can have the rebuilder update both places.

The journal file can be auto-created if absent.

**Comment by sougou** [comment]
@alainjobart @demmer @yuananf @xhh1989 Please provide feedback.

**Comment by yuananf** [comment]
This looks great! The idea of delta change log is genius!
I just have one question:
who is responsible for writing the journal file?
It seems like to be done in any command that changes the keyspace path in global cell?

**Comment by sougou** [comment]
It may have to be done in the functions that update the local cell info.
For tablets, it will be the one that updates the tablet record.
For VSchema, it will likely be the Rebuild command.

**Comment by alainjobart** [comment]
This is a great idea. We only watch one thing, and we get all notifications. A couple things:

1. writing to the change log is not going to be atomic, so you could change a file in topo, then fail writing to the change log, and the change would go undetected. I am not sure how we'd recover from this (apart from restarting all vtgates).

2. I think the vtgates would need to first start watching the log file, then get the values of the subsequent files. To avoid missing updates on startup.

3. It doesn't have to be one log file: could be one entry in a directory (with auto-increment ids in ZK, for instance). And we'd just watch the directory. Maybe if the feature is high level enough in the API, each implementation could implement it the way that works best for them.

4. the data in the payload of these objects is small, might as well save the data in the journal too, and then reading the journal entry is all you need to get the latest value of the data as well. Saves a read.

5. This could also be used to broadcast commands to all vtgates. I don't have a specific use case in mind, apart from re-reading the topology. But that may come in handy at some point.

**Comment by demmer** [comment]
Re 1 and 4, when Sugu and I were talking about this, the idea was to first log a change event in the journal with the time stamp and the path of the object being updated.

Then once successful it would update the object itself with some indication of the change log time stamp to avoid a race where the vtgate sees the journal entry and refreshes the object before the update itself.

**Comment by xhh1989** [comment]
great idea!   I think we does not need to sync all the SrvVSchema to  each local topo. How about sync the SrvSchema information under current cell's keyspaces only?

**Comment by guidoiaquinti** [comment]
> SrvVSchema gets too big: If the number of tablets gets large (>1000), the SrvVSchema file can reach many megabytes, which is something that etcd struggles with.

@sougou I don't think it was mentioned but there's the same (if not worst) limitation in Consul

> Q: What is the per-key value size limitation for Consul's key/value store?
>
> The limit on a key's value size is 512KB. This is strictly enforced and an HTTP 413 status will be returned to any client that attempts to store more than that limit in a value. It should be noted that the Consul key/value store is not designed to be used as a general purpose database. 
>
> see Consul [documentation](https://www.consul.io/docs/faq.html#q-what-is-the-per-key-value-size-limitation-for-consul-39-s-key-value-store-)

**Comment by aquarapid** [comment]
RE large key values, should we consider allowing the vschema to be stored compressed?  I had a large-ish vschema (few hundred kB) lying around that compresses comfortably by 93%...

**Comment by derekperkins** [comment]
The k8s topo already compresses data
